src/GraphCoarsenBuilder.hpp

```diff
@@ -0,0 +1,179 @@
+/*BHEADER**********************************************************************
+ *
+ * Copyright (c) 2018, Lawrence Livermore National Security, LLC.
+ * Produced at the Lawrence Livermore National Laboratory.
+ * LLNL-CODE-745247. All Rights reserved. See file COPYRIGHT for details.
+ *
+ * This file is part of smoothG. For more information and source code
+ * availability, see https://www.github.com/llnl/smoothG.
+ *
+ * smoothG is free software; you can redistribute it and/or modify it under the
+ * terms of the GNU Lesser General Public License (as published by the Free
+ * Software Foundation) version 2.1 dated February 1999.
+ *
+ ***********************************************************************EHEADER*/
+
+/** @file GraphCoarsenBuilder.hpp
+
+    @brief Helper objects for GraphCoarsen::BuildPEdges
+*/
+
+#ifndef __GRAPHCOARSENBUILDER_HPP
+#define __GRAPHCOARSENBUILDER_HPP
+
+#include "smoothG_config.h"
+#include "utilities.hpp"
+#include "mfem.hpp"
+
+namespace smoothg
+{
+
+/**
+   Abstract base class to help building the coarse mass matrix in
+   GraphCoarsen::BuildPEdges()
+*/
+class CoarseMBuilder
+{
+public:
+    virtual ~CoarseMBuilder() {}
+
+    /// The names of the next several methods are not that descriptive or
+    /// informative; they result from removing lines from BuildPEdges()
+    /// and putting it here.
+    virtual void RegisterRow(int agg_index, int row, int cdof_loc, int bubble_counter) = 0;
+
```

Comment:
This function is for setting the values in the (TraceExtension, Bubble)-block (and the transpose) of coarse M. How about call it `SetTraceBubbleBlock`?